### ISSUE TITLE
Identifying post-connection to always-on client

### ISSUE THREAD

**Issue by MystaraTheGreat** [ergochat/ergo#1882]
Hi there,

This is a followup to #1336 and #1319, which both relate to identifying to a nickname after a connection has been made.  I understand from, particularly #1336 that this is difficult owing to the need to tear down an anonymous connection (which is undesirable) or merging two connections (which is complex).  

This is still an issue for me.  Some of our users don't have/can't figure out how to configure SASL or the use of the server password in the form username:password (yeah yeah, I know).  It's very hard to talk a user through this because they need to keep connecting/disconnecting, during which time you can't communicate with them and/or the connection fails.  Sometimes they use a client that you're not familiar with, which makes it near impossible to help locate the relevant fields etc.  

I think, therefore, there is some value in a post-connection identification.

It seems to me that a simple solution is to simply dictate that one of the two connections is the 'master' and then copy that configuration to the other client?  The user themselves could even choose which connection was 'correct'.  This would remove the need for any merging - it would simply be a configuration 'copy'.

**Comment by slingamn** [comment]
It's worth considering.

One issue is synchronizing client state across this operation:

1. Other clients should see a QUIT for the old nickname
2. The initiating client should see a PART for channels it's no longer a member of, and a JOIN for channels that the always-on client is a member of that it wasn't previously JOIN'ed to
3. Probably a few more edge cases like this

**Comment by MystaraTheGreat** [comment]
That all sounds entirely reasonable and exactly how I would expect it to work.

**Comment by yunginnanet** [comment]
It _sounds_ overly-complex to me, but maybe I'm just not visualizing it properly in my head. Where you lost me is the client choosing which config is the proper one to use moving forward.. That's a lot of moving parts in my opinion. Something like that imo would be more suited as a nickserv add-on of some sort rather than an actual feature of the daemon at its core.

**Comment by slingamn** [comment]
I gave this a bit of thought and it's quite difficult to do correctly. Essentially, one has to tear down the old Client object and most of the old Session object, preserving/copying just the `socket`, `capabilities`, `realIP`, and `proxiedIP` members, but while ensuring that no QUIT or ERROR lines are sent to the old socket; this requires creating an alternate code path in `(*Client).destroy`. Then we have to send the necessary state change messages (NICK to communicate the new nickname, JOIN and PART to communicate the new channel memberships) on the socket. Finally, a new Session object has to be created and attached to the relevant Client object. (The recreation of Session is to avoid any data races on `(*Session).socket` or `(*Session).client` by other goroutines who may hold a lingering pointer to the old Session.)

So I'm not ruling this out but I don't think I can deliver it in the near term.

**Comment by MystaraTheGreat** [comment]
Hmmm, I'll defer to your (obvious) expertise here.  But I'm surprised it needs to be that complicated.
Couldn't it be as simple as:

- Join channels that the existing connection is on
- Leave channels that the existing connection is not on
- Kill the old connection
- Change the nickname on the new connection?

Admittedly, I'm not sure I see how the history would work.  But that could presumably be grabbed before the old client is killed and then provided to the new client?

**Comment by slingamn** [comment]
Just to point you a bit in the direction of why it's difficult: "connection" can plausibly refer to either the `Client`, `Session`, or `Socket` object as they appear in the current codebase, and different things need to happen to each of those.

**Comment by MystaraTheGreat** [comment]
Fair enough :)

Thanks for looking at this, by the way.  I'm hopeful it'll appear one day, but I realise it won't be any time soon.

**Comment by slingamn** [comment]
Revisiting this I have an observation that may or may not be helpful :-)

>It's very hard to talk a user through this because they need to keep connecting/disconnecting, during which time you can't communicate with them and/or the connection fails.

Unauthenticated webchat (with Kiwi or Gamja) is very useful for ad-hoc communication in situations like this.

**Comment by cyrneko** [comment]
hm, I am reading through this thread and the other linked issues and I feel this might be why users cannot use `IDENTIFY`... I'm unsure though.

**Comment by slingamn** [comment]
That's correct, if always-on is enabled or there is already a client using the account name as the nickname, this issue will block the use of `NS IDENTIFY` to log into the account.

IMO Ergo really only works well with SASL (or the workaround where you send `PASS accountname:password`). If it's hard or confusing to set up SASL, that's a combination of the following issues:

* Gaps in Ergo's documentation
* Usability issues in clients (SASL PLAIN is quite difficult to configure correctly in Halloy, for example)
* Gaps in the IRCv3 specs that would allow the client to automatically enable SASL (this was one of the goals of [draft/account-registration](https://ircv3.net/specs/extensions/account-registration), but I am not confident that the spec gives clients all the tools they need to support this)

However, I am confident that `NS IDENTIFY` is *not* the answer --- we really need to get the client logged into their account during the handshake, not afterwards :-) Anything else is inefficient, disruptive (produces JOIN/PART and MONITOR spam), and creates correctness issues.

**Comment by cyrneko** [comment]
I've recently set up a network, and in the MOTD put some very basic instructions on registering as well as authenticating, as a good chunk of people on said network haven't used IRC before.

So, it's the easiest and most cross-client way to authenticate, as it doesn't actually require any specific UI to be explained. Just a note about how to send `IDENTIFY`

I guess perhaps I should write a small document about IRC(v3) for said new users...

**Comment by slingamn** [comment]
I would really recommend referring people to the Ergo userguide, which contains our official guidance on client configuration (including SASL), and helps people avoid common pitfalls: https://github.com/ergochat/ergo/blob/master/docs/USERGUIDE.md

**Comment by slingamn** [comment]
If you're telling people to manually send `NS IDENTIFY`, I think this breaks important features of Ergo, such as transparent reconnection (details in the userguide).
